dev/PushNotifications/PushNotificationManager.cpp

```diff
@@ -221,249 +272,371 @@ namespace winrt::Microsoft::Windows::PushNotifications::implementation
-            {
-                THROW_HR_IF_MSG(HRESULT_FROM_WIN32(ERROR_NOT_FOUND), !s_comActivatorRegistration, "ComActivator not registered.");
-                s_comActivatorRegistration.reset();
+                UnregisterSinkHelper();
             }
-
-            if (WI_IsFlagSet(activators, PushNotificationRegistrationActivators::ProtocolActivator))
+            else
             {
-                THROW_HR_IF(HRESULT_FROM_WIN32(ERROR_NOT_FOUND), !s_protocolRegistration);
-                auto coInitialize = wil::CoInitializeEx();
+                if (AppModel::Identity::IsPackagedProcess())
+                {
+                    auto lock{ m_lock.lock_exclusive() };
+                    THROW_HR_IF_MSG(E_UNEXPECTED, !m_comActivatorRegistration, "Need to call Register() before calling Unregister().");
+                    m_comActivatorRegistration.reset();
+                }
 
-                auto notificationPlatform{ PushNotificationHelpers::GetNotificationPlatform() };
+                {
+                    auto lock{ m_lock.lock_exclusive() };
+                    THROW_HR_IF_MSG(E_UNEXPECTED, !m_singletonForegroundRegistration, "Need to call Register() before calling Unregister().");
+                }
 
-                wil::unique_cotaskmem_string processName;
-                THROW_IF_FAILED(GetCurrentProcessPath(processName));
-                LOG_IF_FAILED(notificationPlatform->UnregisterLongRunningActivator(processName.get()));
+                auto notificationsLongRunningPlatform{ PushNotificationHelpers::GetNotificationPlatform() };
+                // Unregister foreground sink with the Long Running Process Singelton
+                THROW_IF_FAILED(notificationsLongRunningPlatform->UnregisterForegroundActivator(m_processName.get()));
 
-                s_protocolRegistration = false;
+                {
+                    auto lock{ m_lock.lock_exclusive() };
+                    m_singletonForegroundRegistration = false;
+                }
             }
         }
         catch (...)
         {
-            PushNotificationTelemetry::ActivatorUnregisteredByApi(wil::ResultFromCaughtException(), activators);
+            PushNotificationTelemetry::ActivatorUnregisteredByApi(wil::ResultFromCaughtException());
             throw;
         }
 
-        PushNotificationTelemetry::ActivatorUnregisteredByApi(S_OK, activators);
+        PushNotificationTelemetry::ActivatorUnregisteredByApi(S_OK);
     }
 
-    void PushNotificationManager::UnregisterAllActivators()
+    void PushNotificationManager::UnregisterAll()
     {
-        THROW_HR_IF(E_NOTIMPL, !::Microsoft::Windows::PushNotifications::Feature_PushNotifications::IsEnabled());
+        bool unregisterCompleted{ false };
+        {
+            auto lock{ m_lock.lock_exclusive() };
+            unregisterCompleted = (!m_singletonForegroundRegistration || !m_comActivatorRegistration);
```

Comment:
Can you please help me to understand this OR condition? for  PackagedAppScenario, in unregister(), looks like we do not change m_singletonForegroundRegistration at all, if  m_singletonForegroundRegistration is always false for packaged AppScenario, looks like unregisterCompleted will always be true here.

Reply:
A lot of this logic needs to be simplified. Unregister seems to destroy state in its current form due to a bug. This check isn't clear at all. Lets not do this. Prevent races between REgister, Unregister and UnregisterAll by using a single bool.

And have simple bool checks to ensure that Unregister is done via THROW_IF checks

Reply:
It should be
if (m_singletonForegroundregistration || m_comActivationREgistration)
{
    Unregister();
}

Reply:
Since the presence of one of them implies that Unregister wasn't called